### PR TITLE
feat(v2): support rightToc, emoji, slug for docs

### PR DESCRIPTION
**Pull request by endiliey**
## Motivation

- Support emoji. Example: `:dog: :+1:` -> 🐶 👍 
- Support slug (GitHub slug way)
- Support right TOC like in v1
image

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

- Locally and production no errors
- Tests added for `rightToc` works. See snapshots for structure

TODO: 
- rereview & audit code
- fix styling

### CODE REVIEW

File: packages/docusaurus-mdx-loader/src/rightToc/__tests__/fixtures/name-exist.mdx [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export const rightToc = ['replaceMe'];
+
+## Thanos
```
Comment (yangshun) [inline packages/docusaurus-mdx-loader/src/rightToc/__tests__/fixtures/name-exist.mdx]: I like these 🌝

File: packages/docusaurus-mdx-loader/src/rightToc/__tests__/index.test.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  const result = await remark()
+    .use(mdx)
+    .use(plugin, options)
+    .process(file);
+
+  return result.toString();
+};
+
+test('no options', async () => {
+  const options = undefined;
```
Comment (yangshun) [inline packages/docusaurus-mdx-loader/src/rightToc/__tests__/index.test.js]: Seems redundant to declare this. Just call `processFixture('just-content')`?

File: packages/docusaurus-mdx-loader/src/rightToc/__tests__/index.test.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  const result = await processFixture('just-content', options);
+  expect(result).toMatchSnapshot();
```
Comment (yangshun) [inline packages/docusaurus-mdx-loader/src/rightToc/__tests__/index.test.js]: Can we use the inline snapshots?

File: packages/docusaurus/lib/default-theme/DocBody/index.js [hunk 4]
```diff
@@ -13,6 +13,22 @@ import Head from '@docusaurus/Head';
 
 import styles from './styles.module.css';
 
+const Headings = ({headings}) => {
+  if (!headings.length) return null;
+  return (
+    <ul className="contents contents__left-border">
```
Comment (yangshun) [inline packages/docusaurus/lib/default-theme/DocBody/index.js]: I see why the left border is duplicated now - you're using recursion. Leave it like this first I'll fix on Infima end and update here.

Comment (yangshun) [reply packages/docusaurus/lib/default-theme/DocBody/index.js]: Or maybe you can render `contents__left-border` if depth === 1

File: packages/docusaurus/lib/default-theme/DocBody/index.js [hunk 5]
```diff
@@ -22,6 +38,7 @@ function DocBody(props) {
   }, []);
 
   const DocContents = content;
+  console.log(content.rightToc);
```
Comment (yangshun) [inline packages/docusaurus/lib/default-theme/DocBody/index.js]: Remove?

File: packages/docusaurus/lib/default-theme/DocBody/index.js (hunk 4 above)
Comment (endiliey) [reply packages/docusaurus/lib/default-theme/DocBody/index.js]: I removed `depth` and set a `isChild` props. Anyway, Its fixed now and will be waiting for infima update

**Comment by endiliey** [comment]
I had to use babel because technically what this is doing is not just MDAST, but MDXAST (MDAST + JSX). 
The babel traverse is just to check if there already exist some statement in markdown like `export const rightToc` so that I can replace it. Otherwise there will be double statement.

**Comment by endiliey** [comment]
Startup time still very fast !! hehe

image

**Pull request merged**